bagel/utility.py

```diff
@@ -1,49 +1,78 @@
-import json
 from pathlib import Path
+from typing import Iterable
 
+import pydantic
 import typer
+from pydantic import ValidationError
 
+from bagel import models
+from bagel.mappings import ALL_NAMESPACES, NB
 
-def file_encoding_error_message(input_p: Path) -> str:
-    """Return a message for when a file cannot be read due to encoding issues."""
-    return typer.style(
-        f"Failed to decode the input file {input_p}. "
-        "Please ensure that both your phenotypic .tsv file and .json data dictionary have UTF-8 encoding.\n"
-        "Tip: Need help converting your file? Try a tool like iconv (http://linux.die.net/man/1/iconv) or https://www.freeformatter.com/convert-file-encoding.html.",
-        fg=typer.colors.RED,
-    )
 
+def generate_context():
+    # Direct copy of the dandi-schema context generation function
+    # https://github.com/dandi/dandi-schema/blob/c616d87eaae8869770df0cb5405c24afdb9db096/dandischema/metadata.py
+    field_preamble = {
+        namespace.pf: namespace.url for namespace in ALL_NAMESPACES
+    }
+    fields = {}
+    for val in dir(models):
+        klass = getattr(models, val)
+        if not isinstance(klass, pydantic.main.ModelMetaclass):
+            continue
+        fields[klass.__name__] = f"{NB.pf}:{klass.__name__}"
+        for name, field in klass.__fields__.items():
+            if name == "schemaKey":
+                fields[name] = "@type"
+            elif name == "identifier":
+                fields[name] = "@id"
+            elif name not in fields:
+                fields[name] = {"@id": f"{NB.pf}:{name}"}
 
-def load_json(input_p: Path) -> dict:
-    """Load a user-specified json type file."""
+    field_preamble.update(**fields)
+
+    return {"@context": field_preamble}
+
+
+def get_subs_missing_from_pheno_data(
+    subjects: Iterable, pheno_subjects: Iterable
+) -> list:
+    """Check a list of subject IDs and return any not found in the provided phenotypic subject list."""
+    return list(set(subjects).difference(pheno_subjects))
+
+
+def extract_and_validate_jsonld_dataset(
+    jsonld: dict, file_path: Path
+) -> models.Dataset:
+    """
+    Strip the context from a user-provided JSONLD and validate the remaining contents
```

Comment:
why do you need a path and a jsonld dictionary input here? Why can't this function just call the `load_json` function directly with the Path? it seems the only use of `jsonld` is to be fed to this function here.

Reply:
good call!